### ISSUE TITLE
Break large modules to improve build times?

### ISSUE THREAD

**Issue by harendra-kumar** [composewell/streamly#433]
The `StreamD` module and `Streamly.Internal.Prelude` modules take a long time to build because these are quite large. We can break them into smaller modules so that the modules can be compiled in parallel and hopefully in lesser time even when building serially (not sure about this).

The `Prelude` and `StreamD` modules can be split into modules based on `generation`, `transformation` and `elimination` APIs. Any APIs common to these modules can be pushed to a core modules e.g. to `Internal.Data.Stream.StreamD.Types` or `Internal.Data.Stream.Prelude`.

We can do similar split for benchmarks as well. We need to measure how much we can gain from this.

**Comment by harendra-kumar** [comment]
The `StreamD` module currently takes 2GB memory to build.

**Comment by harendra-kumar** [comment]
We need to break IsStream module as well on the same lines. We need to put `+RTS -M512M -RTS` in the ghc options to make sure that all modules compile in reasonable amount of heap.

**Comment by harendra-kumar** [comment]
This is already done, closing.
